usaspending_api/bulk_download/filestreaming/csv_selection.py

```diff
@@ -0,0 +1,253 @@
+from usaspending_api.download.v2 import download_column_historical_lookups
+
+BUFFER_SIZE = (5 * 1024 ** 2)
+EXCEL_ROW_LIMIT = 1000000
+
+logger = logging.getLogger('console')
+
+
+def update_number_of_columns(row, download_job):
+    if download_job.number_of_columns is None:
+        download_job.number_of_columns = len(row)
+    else:
+        download_job.number_of_columns = max(download_job.number_of_columns,
+                                             len(row))
+
+
+def csv_row_emitter(body, download_job):
+    header_row = True
+    for row in body:
+        string_buffer = io.StringIO()
+        writer = csv.writer(string_buffer)
+        writer.writerow(row)
+        if header_row:
+            update_number_of_columns(row, download_job)
+            header_row = False
+        else:
+            download_job.number_of_rows += 1
+
+        yield string_buffer.getvalue().encode('utf8')
+
+
+class CsvSource:
+    def __init__(self, model_type, file_type):
+        self.model_type = model_type
+        self.file_type = file_type
+        self.human_names = download_column_historical_lookups.human_names[
+            model_type][file_type]
+        self.query_paths = download_column_historical_lookups.query_paths[
+            model_type][file_type]
+        self.queryset = None
+
+    def values(self, header):
+        query_paths = [self.query_paths[hn] for hn in header]
+        return self.queryset.values_list(query_paths).iterator()
+
+    def columns(self, requested):
+        """Given a list of column names requested, returns the ones available in the source"""
+
+        if requested:
+            result = []
+            for column in requested:
+                if column in self.human_names:
+                    result.append(column)
+        else:
+            result = self.human_names
+
+        # remove headers that we don't have a query path for
+        result = [h for h in result if h in self.query_paths]
+
+        return result
+
+    def row_emitter(self, headers_requested):
+        headers = self.columns(headers_requested)
+        # Not yieling headers as the files can be split
+        # yield headers
+        query_paths = [self.query_paths[hn] for hn in headers]
+        return self.queryset.values(*query_paths)
+
+    def toJsonDict(self):
+        json_dict = {
+            'model_type': self.model_type,
+            'file_type': self.file_type,
+            'query': jsonpickle.dumps(self.queryset.query) if self.queryset is not None else None
+        }
+        return json_dict
+
+
+def calculate_number_of_csvs(queryset, limit):
+    # This could also be done with the commented line below but this performed better
+    # return (queryset.count() // limit) + 1
+    csvs = 0
+    reached_limit = False
+    while not reached_limit:
+        csvs += 1
+        try:
+            queryset[limit*csvs]
+        except IndexError:
+            reached_limit = True
+    return csvs
+
+def date_query_fix(query):
```

Comment:
+1
Wish we could add this to Django query printing.
FYI for regex you should be able to do '\d{4}-\d{2}-\d{2}'

Reply:
Yup that still works. Fixed.